### PR TITLE
Allow Map to use i/o objects

### PR DESCRIPTION
**Pull request by dstansby**
xref #2690

### CODE REVIEW

File: sunpy/map/map_factory.py [hunk 1]
```diff
@@ -189,6 +190,11 @@ def _parse_args(self, *args, **kwargs):
 
             arg = args[i]
 
+            # File object
+            if isinstance(arg, io.IOBase):
+                # Extract the name and then rely on the filename logic below
+                arg = arg.name
```
Comment (Cadair) [inline sunpy/map/map_factory.py:196]: If we are doing this, is there really any point?

Comment (dstansby) [reply sunpy/map/map_factory.py:196]: What do you mean by "is there any point"?

Comment (Cadair) [reply sunpy/map/map_factory.py:196]: Well we aren't using the file object, or respecting the context manager. We will be keeping our own file handle open to the fits file once the context manager is closed or once `f.close()` was called, which to me is us outright lying to the user lol

Comment (dstansby) [reply sunpy/map/map_factory.py:196]: Ha, good point... it's going to need a bit more code wrangling then to do it properly.

**Comment by Cadair** [comment]
I do like the idea of doing this though, I just think we need to pass the file handle through the whole stack, which I appreciate is a bigger effort.

**Comment by dstansby** [comment]
Yep 👍

**Pull request closed without merging**